pkg/i2gw/providers/common/utils.go

```diff
@@ -127,6 +127,34 @@ func ToBackendRef(ib networkingv1.IngressBackend, path *field.Path) (*gatewayv1b
 	}, nil
 }
 
+type orderedPathsByMatchGroup struct {
+	keys []pathMatchKey
+	data map[pathMatchKey][]ingressPath
+}
+
+func groupPaths(rules []ingressRule) orderedPathsByMatchGroup {
+	// we use a slice instead of a map to preserve rules order
+	pathsByMatchGroup := orderedPathsByMatchGroup{}
+
+	for i, ir := range rules {
+		for j, path := range ir.rule.HTTP.Paths {
+			if len(pathsByMatchGroup.keys) == 0 {
+				pathsByMatchGroup.keys = []pathMatchKey{}
+			}
+			if len(pathsByMatchGroup.data) == 0 {
+				pathsByMatchGroup.data = map[pathMatchKey][]ingressPath{}
+			}
```

Comment:
```suggestion
	ingressPathsByMatchKey := & orderedIngressPathsByMatchKey{
		keys: []pathMatchKey{},
		data: make(map[pathMatchKey][]ingressPath),
	}

	for i, ir := range rules {
		for j, path := range ir.rule.HTTP.Paths {
```

Reply:
I considered it, but I think that having structures to nil instead of being empty is a better pattern in case no rule exists. From a functional point of view, the approaches are equivalent.

Reply:
why a better pattern? I think it makes the code less readable and adds additional two if statements.  
memory wise, it is not really an optimization and it is anyway cleaned when the function finishes.

Reply:
I think this is the field of personal opinions. I prefer one way over another, but there is no difference functionally and performance-wise. If you prefer that way I'm ok with changing it.